### DISCUSSION TITLE
Stepper Component - Feature request

### DISCUSSION THREAD

**Discussion by Noah-Haf** [shadcn-ui/ui#1422] [Ideas]
I think a Stepper would be really cool for shadcn UI, something like: 

![image](https://github.com/shadcn-ui/ui/assets/61884850/f73dfe48-1622-4c36-bd76-4d35bddec671)

**Comment by kelvindesman** [comment]
check out this pr https://github.com/shadcn-ui/ui/pull/318

**Reply by guifeliper** [reply]
Do you know how can I use it, without being merged?

**Reply by papsavas** [reply]
Just copy `stepper.tsx` & `useStepper.ts` files

**Reply by Mahizha-N-S** [reply]
i cant find the files

**Reply by papsavas** [reply]
> i cant find the files

The PR has changed alot since last year 😅
You can find the files in the [author's repo](https://github.com/damianricobelli/shadcn-stepper)
And some examples [here](https://shadcn-stepper.vercel.app)

> Unfortunately the examples do not include the documentation, but you can find the code inside the repo [examples folder](https://github.com/damianricobelli/shadcn-stepper/tree/main/examples)

**Reply by Mahizha-N-S** [reply]
i get that,
But which files should we copy and paste and how to use it?
like in that, i have this
import { Step, type StepItem, Stepper, useStepper } from "@/components/stepper";
but stepper is a folder right

**Comment by mikdanjey** [comment]
@shadcn, We need Wizard / Stepper this feature.

**Comment by dorji-dev** [comment]
Discussion link to which I have originally posted this: [https://github.com/shadcn-ui/ui/discussions/1869](https://github.com/shadcn-ui/ui/discussions/1869)

For anyone reaching till this point, I am doing some research for my company on which form library to choose: formik or react-hook-form. For that I had to create a multi-step form for both so that we can compare DX and features provided by each library. Done with both forms.

Features include: 

- Validation before next step
- Uses controlled component as well with shadcn select and single datepicker
- Conditional validation for some inputs based the value of another input
- If there is error on submit from api, and if the error message contains the field name as well with which the error is associated to, then the step that contains that field is made active and also the input is focused unless it is a controlled component
- If the api error doesn't have field name then a generic message is shown which persists across step change and is being reset automatically on next submit.

All those are done with the help of hooks and components like `setError`, `FormProvider`, and `useFormContext` exposed by the `react-hook-form` and was a joy to create the stepper. Didn't need any global state manager/context. In the end on submit, you get all the data as all the steps are inside the `FormProvider`. 

Used NextJs app router and here is the link to the page: https://github.com/dorji-dev/formik-vs-react-hook-form/blob/main/src/app/hook-multi-step/page.tsx

**You can find the instructions on how to integrate it with an existing project via the repo README file**

A preview: 

<img width="901" alt="Screenshot 2024-06-14 at 15 47 39" src="https://github.com/shadcn-ui/ui/assets/139944071/b3ef7e9e-983c-4a51-92be-40d3c15c234c">

**Reply by joshwisehub** [reply]
just awesome i have intergrate it  in my react project using tanstack router and it worked smoothly.

**Comment by HassanDev13** [comment]
Hello everyone, here a small try from me : 

https://github.com/HassanDev13/stepper-react/

**Comment by Noah-Haf** [comment]
[Looks Like it's coming! ](https://github.com/shadcn-ui/ui/pull/318#issuecomment-2623740276)

**Comment by manhtrv** [comment]
https://www.shadcn-vue.com/docs/components/stepper
please fork it from shadcn-vue
